### PR TITLE
auth: add application default credentials as fallback

### PR DESCRIPTION
**Pull request by vikkyrk**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by vikkyrk** [comment]
R: @charlesccychen @aaltay

**Comment by lukecwik** [comment]
fyi, the java SDK dropped explicit support for the additional flags like service account name, service account key file and always just defaults to application default credentials.

**Comment by vikkyrk** [comment]
@lukecwik thanks, yes we have cleanup planned for this pretty soon. Right now this PR unblocks datastoreio.

File: sdks/python/apache_beam/internal/auth.py [hunk 1]
```diff
@@ -163,4 +164,20 @@ def get_service_credentials():
...
+        logging.warning("Unable to find credentials from gcloud.")
+
+      # Falling back to application default credentials.
+      try:
+        credentials = GoogleCredentials.get_application_default()
+        credentials = credentials.create_scoped(client_scopes)
+        logging.info('Connecting using Google Application Default Credentials.')
+        return credentials
+      except Exception:
+        logging.exception("Unable to find default credentials to use.")
```
Comment (aaltay) [inline sdks/python/apache_beam/internal/auth.py]: Should you return something in this case, or raise an exception?

File: sdks/python/apache_beam/internal/auth.py [hunk 2]
```diff
@@ -163,4 +164,20 @@ def get_service_credentials():
...
 
     else:
-      return _GCloudWrapperCredentials(user_agent)
+      try:
+        credentials = _GCloudWrapperCredentials(user_agent)
+        # Check if we are able to get an access token. If not fallback to
+        # application default credentials.
+        credentials.get_access_token()
+        return credentials
+      except Exception:
```
Comment (aaltay) [inline sdks/python/apache_beam/internal/auth.py]: Do not use a broad exception. what is the expected `Exception` type here?

File: sdks/python/apache_beam/internal/auth.py [hunk 3]
```diff
@@ -163,4 +164,20 @@ def get_service_credentials():
...
+        logging.warning("Unable to find credentials from gcloud.")
+
+      # Falling back to application default credentials.
+      try:
+        credentials = GoogleCredentials.get_application_default()
+        credentials = credentials.create_scoped(client_scopes)
+        logging.info('Connecting using Google Application Default Credentials.')
```
Comment (aaltay) [inline sdks/python/apache_beam/internal/auth.py]: Do you need this log? The other two warning logs have enough information.

File: sdks/python/apache_beam/internal/auth.py (hunk 2 above)
Comment (vikkyrk) [inline sdks/python/apache_beam/internal/auth.py]: Fixed

File: sdks/python/apache_beam/internal/auth.py (hunk 3 above)
Comment (vikkyrk) [inline sdks/python/apache_beam/internal/auth.py]: Made it debug

File: sdks/python/apache_beam/internal/auth.py (hunk 1 above)
Comment (vikkyrk) [inline sdks/python/apache_beam/internal/auth.py]: Fixed

**Comment by aaltay** [comment]
Thank you, LGTM.

File: sdks/python/apache_beam/internal/auth.py [hunk 4]
```diff
@@ -163,4 +164,22 @@ def get_service_credentials():
...
     else:
-      return _GCloudWrapperCredentials(user_agent)
+      try:
+        credentials = _GCloudWrapperCredentials(user_agent)
+        # Check if we are able to get an access token. If not fallback to
+        # application default credentials.
+        credentials.get_access_token()
+        return credentials
+      except AuthenticationException:
+        logging.warning("Unable to find credentials from gcloud.")
```
Comment (aaltay) [inline sdks/python/apache_beam/internal/auth.py]: I believe rest of the sdk uses `'` single quotes for log statements, please change them.

**Comment by aaltay** [comment]
Thank you, LGTM. (Please make sure at least some tests pass. Jenkins did not run, Travis failed for some other reason seemingly.)

**Comment by vikkyrk** [comment]
It did run the python tests if you look at the full log. For some reason it shows didn't run for all the java targets.

**Comment by vikkyrk** [comment]
R: @robertwb @charlesccychen

**Comment by vikkyrk** [comment]
@aaltay Now that we have application default cred, just using that in datastoreio instead of the googledatastore one.

**Comment by aaltay** [comment]
LGTM

**Pull request closed without merging**